module/zfs/arc.c

```diff
@@ -8101,6 +8125,17 @@ l2arc_evict(l2arc_dev_t *dev, uint64_t distance, boolean_t all)
 		mutex_exit(hash_lock);
 	}
 	mutex_exit(&dev->l2ad_mtx);
+
+out:
+	if (rerun) {
+		/*
+		 * Bump device hand to the device start if it is approaching the
+		 * end. l2arc_evict() has already evicted ahead for this case.
+		 */
+		dev->l2ad_hand = dev->l2ad_start;
+		dev->l2ad_first = B_FALSE;
+		l2arc_evict(dev, distance, B_FALSE);
```

Comment:
I think doing this iteratively would be _slightly_ preferable to minimize stack depth.   Of course, In this case the stack frame is small, and the stack depth will always be shallow so in practice I can't see this ever being an issue.  So I'm OK with it either way.

Reply:
Agreed, I will change it to iteration, seems trivial to do so.